model/test_results.go

```diff
@@ -378,3 +334,108 @@ type TestResult struct {
 	TestStartTime  time.Time `bson:"test_start_time"`
 	TestEndTime    time.Time `bson:"test_end_time"`
 }
+
+// TestResultsFindOptions allows for querying for test results with or without
+// an execution value.
+type TestResultsFindOptions struct {
+	TaskID          string
+	DisplayTaskName string
+	Execution       int
+	EmptyExecution  bool
+}
+
+func (opts *TestResultsFindOptions) createFindOptions() *options.FindOptions {
+	findOpts := options.Find().SetSort(bson.D{{Key: bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoExecutionKey), Value: -1}})
+	if opts.TaskID != "" {
+		findOpts = findOpts.SetLimit(1)
+	}
+
+	return findOpts
+}
+
+func (opts *TestResultsFindOptions) createFindQuery() map[string]interface{} {
+	search := bson.M{}
+	if opts.TaskID != "" {
+		search[bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoTaskIDKey)] = opts.TaskID
+	} else {
+		search[bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoDisplayTaskNameKey)] = opts.DisplayTaskName
+	}
+	if !opts.EmptyExecution {
+		search[bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoExecutionKey)] = opts.Execution
+	}
+
+	return search
+}
+
+func (opts *TestResultsFindOptions) createErrorMessage() string {
+	var msg string
+	if opts.TaskID != "" {
+		msg = fmt.Sprintf("could not find test result record with task_id %s", opts.TaskID)
+	} else {
+		msg = fmt.Sprintf("could not find test results records with display task name %s", opts.DisplayTaskName)
+	}
+	if !opts.EmptyExecution {
+		msg += fmt.Sprintf(" and execution %d", opts.Execution)
+	}
+	msg += " in the database"
+
+	return msg
+}
+
+// FindTestResults searches the database for the TestResults associated with
+// the provided options. The environment should not be nil. If execution is
+// empty, it will default to the most recent execution.
+func FindTestResults(ctx context.Context, env cedar.Environment, opts TestResultsFindOptions) ([]TestResults, error) {
+	if env == nil {
+		return nil, errors.New("cannot find with a nil environment")
+	}
+
+	if opts.TaskID == "" && opts.DisplayTaskName == "" {
+		return nil, errors.New("cannot find without a task_id or display_task_name")
+	}
+
+	results := []TestResults{}
+	cur, err := env.GetDB().Collection(testResultsCollection).Find(ctx, opts.createFindQuery(), opts.createFindOptions())
+	if err != nil {
+		return nil, errors.Wrap(err, "finding test results record(s)")
+	}
+	if err := cur.All(ctx, &results); err != nil {
+		return nil, errors.Wrap(err, "decoding test result record(s)")
+	}
+	if len(results) == 0 {
+		return nil, errors.Wrap(mongo.ErrNoDocuments, opts.createErrorMessage())
+	}
+
+	execution := results[0].Info.Execution
```

Comment:
Should we wrap this block with a check for `if opts.EmptyExecution`, since it should only be necessary when execution is not specified?

Reply:
We still need to fill set `populated` and `env` either way.